crates/storage/provider/src/providers/snapshot/writer.rs

```diff
@@ -0,0 +1,268 @@
+                    let (writer, data_path) = Self::open(
+                        segment,
+                        self.writer.user_header().block_start() - 1,
+                        self.reader.clone(),
+                    )?;
+                    self.writer = writer;
+                    self.data_path = data_path;
+
+                    NippyJar::<SegmentHeader>::load(&previous_snap)?.delete()?;
+                } else {
+                    // Update `SegmentHeader`
+                    self.writer.user_header_mut().prune(num_rows);
+                }
+
+                num_rows -= len;
+            } else {
+                // Update `SegmentHeader`
+                self.writer.user_header_mut().prune(num_rows);
+
+                // Truncate data
+                self.writer.prune_rows(num_rows as usize)?;
+                num_rows = 0;
+            }
+        }
+
+        // Only Transactions and Receipts
+        if let Some(last_block) = last_block {
+            let header = self.writer.user_header_mut();
+            header.set_block_range(header.block_start()..=last_block);
+        }
+
+        // Commits new changes to disk.
+        self.commit()?;
+
+        Ok(())
+    }
+
+    /// Appends column to snapshot file.
+    fn append_column<T: Compact>(&mut self, column: T) -> ProviderResult<()> {
+        self.buf.clear();
+        column.to_compact(&mut self.buf);
+
+        self.writer.append_column(Some(Ok(&self.buf)))?;
+        Ok(())
+    }
+
+    /// Appends to tx number-based snapshot file.
+    ///
+    /// Returns the current [`TxNumber`] as seen in the static file.
+    fn append_with_tx_number<V: Compact>(
+        &mut self,
+        segment: SnapshotSegment,
+        tx_num: TxNumber,
+        value: V,
+    ) -> ProviderResult<TxNumber> {
+        debug_assert!(self.writer.user_header().segment() == segment);
+
+        if self.writer.user_header().tx_range().is_none() {
+            self.writer.user_header_mut().set_tx_range(tx_num..=tx_num);
+        } else {
+            self.writer.user_header_mut().increment_tx();
+        }
+
+        self.append_column(value)?;
+
+        Ok(self.writer.user_header().tx_end())
+    }
+
+    /// Appends transaction to snapshot file.
+    ///
+    /// Returns the current [`TxNumber`] as seen in the static file.
+    pub fn append_transaction(
+        &mut self,
+        tx_num: TxNumber,
+        tx: TransactionSignedNoHash,
+    ) -> ProviderResult<TxNumber> {
+        self.append_with_tx_number(SnapshotSegment::Transactions, tx_num, tx)
+    }
+
+    /// Appends receipt to snapshot file.
+    ///
+    /// Returns the current [`TxNumber`] as seen in the static file.
+    pub fn append_receipt(
+        &mut self,
+        tx_num: TxNumber,
+        receipt: Receipt,
+    ) -> ProviderResult<TxNumber> {
+        self.append_with_tx_number(SnapshotSegment::Receipts, tx_num, receipt)
+    }
+
+    /// Prunes `to_delete` number of transactions from snapshots.
```

Comment:
can we add more context what this actually does?

not obvious what prune here means exactly